qa/rpc-tests/cfund-rawtx-proposal-vote.py

```diff
@@ -0,0 +1,200 @@
+
+        assert (self.nodes[0].getproposal(proposalid1)['votesYes'] == 1)
+        assert (self.nodes[0].getproposal(proposalid1)['votesNo'] == 1)
+
+        # Insert bad vote tx with double vote in string
+        pr0_bad_vote_tx = self.create_vote_tx('c3', 'c4c4', proposalid0)
+        self.nodes[0].coinbaseoutputs([pr0_bad_vote_tx])
+        slow_gen(self.nodes[0], 1)
+        assert (self.nodes[0].getproposal(proposalid0)['votesYes'] == 4)
+
+        # vote to accept proposal0
+        self.vote_to_new_cycle(prop0_vote_tx_yes)
+        assert (self.nodes[0].getproposal(proposalid0)['status'] == 'accepted')
+        assert (self.nodes[0].getproposal(proposalid0)['stateChangedOnBlock'])
+
+        # vote to reject proposal 1
+        self.vote_to_new_cycle(prop1_vote_tx_no)
+        print(self.nodes[0].getproposal(proposalid1))
+        assert (self.nodes[0].getproposal(proposalid1)['status'] == 'rejected')
+
+
+    def vote_to_new_cycle(self, voteTx):
+        # Move one past the end of the cycle
+
+        blocksToEnd = self.nodes[0].cfundstats()["votingPeriod"]["ending"] - self.nodes[0].cfundstats()["votingPeriod"]["current"] + 1
+
+        for i in range(0, blocksToEnd):
+            self.nodes[0].coinbaseoutputs([voteTx])
+            slow_gen(self.nodes[0], 1)
+
+
+    def get_transaction_from_block(self, blochHash):
+
+        block = self.nodes[0].getblock(blochHash)
+
+        tx = self.nodes[0].gettransaction(block['tx'][0])
+
+        decodedTx = self.nodes[0].decoderawtransaction(tx['hex'])
+
+        return decodedTx
+
+    def is_vote_hash_in_vout(self, decodedTx, voteHash):
+
+        vout = decodedTx['vout']
+
+        isInVout = False
+        for tx in vout:
+            if tx['scriptPubKey']['hex'] == voteHash:
+                isInVout = True
+
+        return isInVout
+
+    def reverse_byte_str(self, hex_str):
+        return ''.join([c for t in zip(hex_str[-2::-2], hex_str[::-2]) for c in t])
+
+    def create_vote_tx(self, vote_type, vote, p_hash):
+        """
+        Creates voting hex to be included into the coinbase.
+        Args:
+            vote_type: proposal:'c2', payment request: 'c3'
+            vote: yes: 'c4', no:'c5'
+            p_hash: hash of the proposal/payment request
+
+        Returns:
+            str: hex data to include into the coinbase
+        """
+        # Byte-reverse hash
+        reversed_hash = self.reverse_byte_str(p_hash)
+
+        # Create voting string
+        vote_str = '6a' + 'c1' + vote_type + vote + '20' + reversed_hash
+
+        # Create raw vote tx
+        vote_tx = self.nodes[0].createrawtransaction(
+            [],
+            {vote_str: 0},
+            "", 0
+        )
+        return vote_tx
+
+    def start_new_cycle(self):
+        # Move to the end of the cycle
+        slow_gen(self.nodes[0], self.nodes[0].cfundstats()["votingPeriod"]["ending"] - self.nodes[0].cfundstats()["votingPeriod"]["current"])
+
+    def activate_cfund(self):
```

Comment:
Should we pull these functions out into a utils lib? We use them in all the cfund tests right?
I'm sure there are some other functions that we have duplicates of in our other tests too

Reply:
Yes and no, there are some duplicates, and some have been moved out, but it one of those time vs reward situations .